packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,205 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    getMarkLength,
+    isAttachGroup,
+    isDetachMark,
+    isReattach,
+    MarkListFactory,
+    splitMark,
+    Transposed as T,
+} from "../../changeset";
+import { clone, fail } from "../../util";
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Composes a sequence of changesets into a single changeset.
+ * @param changes - The changesets to be applied.
+ * Each changeset in the list is assumed to be applicable after the previous one.
+ * @returns A changeset that is equivalent to applying each of the given `changes` in order.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Tombstone information is ignored.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function compose(changes: SequenceChangeset[]): SequenceChangeset {
+    const base: SequenceChangeset = {
+        marks: {},
+    };
+    for (const change of changes) {
+        foldInFieldMarks(change.marks, base.marks);
+    }
+    return base;
+}
+
+function foldInFieldMarks(newFieldMarks: T.FieldMarks, baseFieldMarks: T.FieldMarks) {
+    for (const key of Object.keys(newFieldMarks)) {
+        const newMarkList = newFieldMarks[key];
+        baseFieldMarks[key] ??= [];
+        const baseMarkList = foldInMarkList(newMarkList, baseFieldMarks[key]);
+        baseFieldMarks[key] = baseMarkList;
+        while (typeof baseMarkList[baseMarkList.length - 1] === "number") {
+            baseMarkList.pop();
+        }
+        if (baseMarkList.length === 0) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+            delete baseFieldMarks[key];
+        }
+    }
+}
+
+function foldInMarkList(
+    newMarkList: T.MarkList,
+    baseMarkList: T.MarkList,
+): T.MarkList {
+    const factory = new MarkListFactory();
+    let iBase = 0;
+    let iNew = 0;
+    let nextBaseMark: T.Mark | undefined = baseMarkList[iBase];
+    let nextNewMark: T.Mark | undefined = newMarkList[iNew];
+    while (nextNewMark !== undefined) {
```

Comment:
It would be nice if we could share extract some logic for iterating pairs of mark lists and share it with `rebase`. It might not be easy since in one case both lists refer to the same input tree, and in the other case one list refers to the output tree of the other list. But a cursory glance at `rebase` looks like it has some very similar iteration code.

Reply:
Done with `StackyIterator`